pkg/peertls/io_util.go

```diff
@@ -28,11 +27,7 @@ func writeCerts(certs [][]byte, path string) error {
 		return errs.New("unable to open file \"%s\" for writing", path, err)
 	}
 
-	defer func() {
-		if err := file.Close(); err != nil {
-			log.Printf("Failed to close file: %s\n", err)
-		}
-	}()
+	defer func() { _ = file.Close() }()
```

Comment:
why are you ignoring the error ?

Reply:
Because file.Close is called later in the function (as it should be) and during successful operation this will always be a double close error.

Reply:
if we are explicitly calling close and should be, should we be defering close ?

Reply:
yeah, what if we return early due to one of the other errors? we don't want to leak the file handle

Reply:
Oh, we should always be explicit with closing, especially file handles. If you don't explicitly close a file handle, the last buffered data might fail to sync to disk and you might not know. But that said on an unhappy path it's also good to confirm that we don't leak anything

Reply:
explicit meaning we call close explicitly before returning instead of just defer'ing the close.

Reply:
If we have a derfer'd close, I don't think we _should_ be called later. Calling close twice seems to me, to not be something we want, and ignoring the error seems to just be brushing off the fact we aren't doing the correct thing.

Reply:
> explicit meaning we call close explicitly before returning instead of just defer'ing the close.
right
> If we have a derfer'd close, I don't think it should be called later. Calling close twice seems to me, to not be something we want, and ignoring the error seems to just be brushing off the fact we aren't doing the correct thing.

okay so there's two competing concerns. 

the first concern is if you write stuff to a file, on certain operating systems, the writes may not have successfully made it to disk unless close returns with no error. so to be sure you wrote a file, you need to explicitly close and check the error, and if there was an error on closing, your operation may have failed.

the second concern is resource leaks. if you forget to close the file in an error case, you leak a file handle.

the way i've seen this work most commonly is explicit closing and deferred closing together. most close methods are written to be okay with being called twice. 

is there a better pattern?

Reply:
Cool I hadn't realized that a file write might not return the error until close. Love learning new things. 

Re: better pattern, the os package https://golang.org/pkg/os/ returns an `ErrClosed` if attempting to close an already closed file. I would recommend not ignoring the entire error but ignoring the specific error we don't care about so `if err != nil && err != os.ErrClosed {`

Reply:
that's a great idea, i'll do that